source/class/qx/data/Array.js

```diff
@@ -31,6 +31,51 @@ qx.Class.define("qx.data.Array", {
   include: qx.data.marshal.MEventBubbling,
   implement: [qx.data.IListData],
 
+  /**
+   * Allow a `qx.data.Array` instance to be indexed as if it were an ordinary
+   * array. Instead of using the provided API to retrieve an indexed item
+   * (`getItem()`), store a value in an indexed item (`setItem()`), or remove
+   * an indexed item (`remove()`), this delegate allows thsoe operations to
+   * use normal array operators, e.g., `value = arr[2]`, `arr[2] = 42;`,
+   * `delete arr[2]`.
+   *
+   * @EXPERIMENTAL
+   */
+  delegate: {
+    get(property) {
+      // If the property is a number or string representing a number...
+      if (typeof property != "symbol" && +property === +property) {
+        // ... then use the method for retrieving an item
+        return this.getItem(property);
+      } else {
+        // otherwise, use the default action
+        return this[property];
+      }
+    },
+
+    set(property, value) {
+      // If the property is a number or string representing a number...
+      if (typeof property != "symbol" && +property === +property) {
+        // ... then use the method for modifying an item
+        this.setItem(property, value);
+      } else {
+        // otherwise, use the default action
+        this[property] = value;
+      }
+    },
+
+    delete(property) {
+      // If the property is a number or string representing a number...
```

Comment:
`delete dataArray[2]` is not the same as `dataArray.removeAt(2)` because `delete` will simply create a hole in the array where as `removeAt` will shuffle the elements down.  Does qx.data.Array support sparse arrays?

Reply:
`delete dataArray[2]` in qx.data.Array is implemented as `dataArray.removeAt(2)`. I don't believe that qx.data.Array supports sparse arrays, so we probably don't want `dataArray.removeAt(2)` to leave an `undefined` hole. The implementation is trivially changeable, though. Do you have suggestions for a different implementation? (It's marked as experimental, so what we do now can be changed later.)

Reply:
But it's different semantics - calling `delete myNativeArray[2]` is not the same as `myDataArray.removeAt(2)`

Reply:
Understood. What would you have it do instead?

Reply:
throw an exception?

Reply:
Ok. It now throws an exception by default. If environment variable `qx.data.Array.deleteAsRemoveAt` is `true` then it acts as I had implemented it originally.